pkg/endpoint/public_ip_test.go

```diff
@@ -0,0 +1,188 @@
+			Namespace: testNamespace,
+		}
+
+		backendConfig = map[string]string{}
+	})
+
+	When("a LoadBalancer with Ingress IP is specified", func() {
+		It("should return the IP", func() {
+			backendConfig[publicIPConfig] = "lb:" + testServiceName
+			client := fake.NewSimpleClientset(serviceWithIngress(v1.LoadBalancerIngress{Hostname: "", IP: testIP}))
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ip).To(Equal(testIP))
+		})
+	})
+
+	When("a LoadBalancer with Ingress hostname is specified", func() {
+		It("should return the IP", func() {
+			backendConfig[publicIPConfig] = "lb:" + testServiceName
+			client := fake.NewSimpleClientset(serviceWithIngress(v1.LoadBalancerIngress{Hostname: testIPDNS + ".nip.io",
+				IP: ""}))
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ip).To(Equal(testIPDNS))
+		})
+	})
+
+	When("a LoadBalancer with no ingress is specified", func() {
+		It("should return error", func() {
+			backendConfig[publicIPConfig] = "lb:" + testServiceName
+			client := fake.NewSimpleClientset(serviceWithIngress())
+			_, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	When("an IPv4 entry specified", func() {
+		It("should return the IP", func() {
+			backendConfig[publicIPConfig] = "ipv4:" + testIP
+			client := fake.NewSimpleClientset()
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ip).To(Equal(testIP))
+		})
+	})
+
+	When("a DNS entry specified", func() {
+		It("should return the IP", func() {
+			backendConfig[publicIPConfig] = "dns:" + testIPDNS + ".nip.io"
+			client := fake.NewSimpleClientset()
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ip).To(Equal(testIPDNS))
+		})
+	})
+
+	When("an API entry specified with api.ipify.org", func() {
+		It("should return some IP", func() {
+			backendConfig[publicIPConfig] = "api:api.ipify.org"
+			client := fake.NewSimpleClientset()
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(net.ParseIP(ip)).NotTo(BeNil())
+		})
+	})
+
+	When("an API entry specified with api.my-ip.io/ip", func() {
+		It("should return some IP", func() {
+			backendConfig[publicIPConfig] = "api:api.my-ip.io/ip"
+			client := fake.NewSimpleClientset()
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(net.ParseIP(ip)).NotTo(BeNil())
+		})
+	})
+
+	When("an API entry specified with ip4.seeip.org", func() {
+		It("should return some IP", func() {
+			backendConfig[publicIPConfig] = "api:ip4.seeip.org"
+			client := fake.NewSimpleClientset()
+			ip, err := getPublicIP(submSpec, client, backendConfig)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(net.ParseIP(ip)).NotTo(BeNil())
+		})
+	})
+
```

Comment:
I'm not sure we should test these specific sites - we may get intermittent failures.  The main thing is to test the algorithm. We could start a local HTTP server for the test (on `localhost:<some port>`) and use that URL. If we want to verify these specific sites for validity as defaults then we can add a periodic test.

Reply:
I think we really should. Probably unit test is not the best place to test it.
But we definitely should test them periodically to make sure that the defaults we offer to customers are functional and healthy.

I can remove those tests if we agree to enable some periodic check for those resolvers in CI

Reply:
I agree we need to test the remote resolvers, not to do their monitoring for them, but just to know that they are functional as you say. However their failure should never create difficulty in evaluating or merging a PR. So a periodic test is fine for this.

Reply:
Thanks for your feedback, I agree it will be better to periodically monitor. I will create a separate PR for that, some github action + bash should suffice.